onnxruntime/contrib_ops/cpu/gather_nd.cc

```diff
@@ -45,12 +46,16 @@ Status GatherNDBase::PrepareForCompute(OpKernelContext* context, Prepare& p) con
   auto output_tensor = context->Output(0,TensorShape(shape));
   std::vector<int64_t> element_counts(last_indice_dimension, 0LL); // Number of elements for each input dimension
 
-#ifdef USE_OPENMP
-#pragma omp parallel for
-#endif
-  for (int64_t i = 0; i < last_indice_dimension; ++i) {
-    element_counts[i] = input_shape.SizeFromDimension(i + 1);
-}
+  auto ctx_internal = static_cast<OpKernelContextInternal*>(context);
+  auto thread_pool = const_cast<concurrency::ThreadPool*>(ctx_internal->GetOperatorThreadPool());
+
+  int64_t shard_size = thread_pool->CalculateShardSize(last_indice_dimension);
```

Comment:
Can ParallelFor itself take care of calculating the shard size instead of relying on every call site to do so? I guess we can expose 2 interfaces - one that takes care of the shard size and one that requires you to pass the shard size (for additional tuning).

Reply:
That is preferable.  There are good uses for dynamically calculating shard sizes based on data that is not generally known to the parallelfor invocation.  I just built the pieces to enable.